### ISSUE TITLE
Get /usr/bin/ld: cannot find -lclntsh while installing oracle_fdw

### ISSUE THREAD

**Issue by MarvinDBA** [laurenz/oracle_fdw#322]
I am trying to install oracle_fdw for PostgreSQL 10.3 to access Oracle RDBMS 19.1.
I have installed with yum the following rpms:

    oracle-instantclient19.3-basic-19.3.0.0.0-1.x86_64.rpm
    oracle-instantclient19.3-devel-19.3.0.0.0-1.x86_64.rpm

After that installation, I tracked where the files were installed:

```
rpm -ql oracle-instantclient18.5-basic-18.5.0.0.0-3 

/usr/lib/oracle/18.5/client64/bin/adrci
/usr/lib/oracle/18.5/client64/bin/genezi
/usr/lib/oracle/18.5/client64/lib/libclntsh.so.18.1
/usr/lib/oracle/18.5/client64/lib/libclntshcore.so.18.1
/usr/lib/oracle/18.5/client64/lib/libipc1.so
/usr/lib/oracle/18.5/client64/lib/libmql1.so
/usr/lib/oracle/18.5/client64/lib/libnnz18.so
/usr/lib/oracle/18.5/client64/lib/libocci.so.18.1
/usr/lib/oracle/18.5/client64/lib/libociei.so
/usr/lib/oracle/18.5/client64/lib/libocijdbc18.so
/usr/lib/oracle/18.5/client64/lib/libons.so
/usr/lib/oracle/18.5/client64/lib/liboramysql18.so
/usr/lib/oracle/18.5/client64/lib/network/admin/README
/usr/lib/oracle/18.5/client64/lib/ojdbc8.jar
/usr/lib/oracle/18.5/client64/lib/xstreams.jar
/usr/share/oracle/18.5/client64/doc/BASIC_LICENSE
/usr/share/oracle/18.5/client64/doc/BASIC_README
```

I have unzipped oracle_fdw at:

    /home/postgres/oracle_fdw/oracle_fdw-2.1.0

I also have:

    echo $ORACLE_HOME
    /usr/share/oracle/19.3/client64/instantclient_19_3
    echo $LD_LIBRARY_PATH
    /usr/lib/oracle/19.3/client64/lib

When I run the make command, I get this error:

```
gcc -Wall -Wmissing-prototypes -Wpointer-arith -Wdeclaration-after-statement -Wendif-labels -Wmissing-format-attribute -Wformat-security -fno-strict-aliasing -fwrapv -fexcess-precision=standard -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches -m64 -mtune=generic -fPIC -shared -o oracle_fdw.so oracle_fdw.o oracle_utils.o oracle_gis.o -L/usr/pgsql-10/lib  -Wl,--as-needed  -L/usr/lib64 -Wl,--as-needed -Wl,-rpath,'/usr/pgsql-10/lib',--enable-new-dtags  -L/usr/share/oracle/19.3/client64/instantclient_19_3 -L/usr/share/oracle/19.3/client64/instantclient_19_3/bin -L/usr/share/oracle/19.3/client64/instantclient_19_3/lib -lclntsh -L/usr/lib/oracle/12.2/client/lib -L/usr/lib/oracle/12.2/client64/lib -L/usr/lib/oracle/12.1/client/lib -L/usr/lib/oracle/12.1/client64/lib -L/usr/lib/oracle/11.2/client/lib -L/usr/lib/oracle/11.2/client64/lib -L/usr/lib/oracle/11.1/client/lib -L/usr/lib/oracle/11.1/client64/lib -L/usr/lib/oracle/10.2.0.5/client/lib -L/usr/lib/oracle/10.2.0.5/client64/lib -L/usr/lib/oracle/10.2.0.4/client/lib -L/usr/lib/oracle/10.2.0.4/client64/lib -L/usr/lib/oracle/10.2.0.3/client/lib -L/usr/lib/oracle/10.2.0.3/client64/lib
/usr/bin/ld: cannot find -lclntsh
collect2: error: ld returned 1 exit status
make: *** [oracle_fdw.so] Error 1
```

There was already a thread created with that error, but I didn't see the solution for it.

Any idea how I can solve it?

Kind Regards

**Comment by laurenz** [comment]
Does the "devel" package contain `libclntsh.so`?
If yes, in which directory?
And where is `oci.h`?

The `Makefile` does not cover the 19.3 RPMs yet; if you tell me the file locations, I can add them.

**Comment by MarvinDBA** [comment]
Thanks for your very fast answer.
In the meantime, I started everything from scrach, and I got a different error.
I used this time the zip files and not the rpms.
I think it is easier if I describe the procedure I followed:


I used Linux x86-64 OEL 7.4, and PostgreSQL 10.3.


1) Download and install files for build:

-> Create directory: 
oracle_fdw

-> Will be use store and install - locally - the oracle_fdw conribution and the Oracle Instant Client libraries.

-> Create subdirectory:
oracle-client

-> Download the Oracle Instant Client libraries:

OTN -> Instant Client Downloads for Linux x86-64 (64-bit):

- instantclient-sdk-linux.x64-19.3.0.0.0dbru.zip
- instantclient-basic-linux.x64-19.3.0.0.0dbru.zip

-> Download these zip files to oracle_fdw/oracle-client/

-> Download the last available oracle_fdw:

https://pgxn.org/dist/oracle_fdw/

- oracle_fdw-2.1.0.zip

-> Download this zip file in oracle_fdw/


2) Prepare Oracle Instant Client libraries:

-> In oracle_fdw/oracle-client unzip instantclient-basic-linux.x64-19.3.0.0.0dbru.zip
-> Then, still in oracle_fdw/oracle-client unzip instantclient-sdk-linux.x64-19.3.0.0.0dbru.zip
-> subdirectory instantclient_19_3 is created:

oracle_fdw/oracle-client/instantclient_19_3


3) Prepare oracle_fdw:

-> Ini oracle_fdw/ unzip oracle_fdw-2.1.0.zip


4) Install postgresql-contrib (should be already installed):

sudo yum install postgresql-contrib


5) Install postgresql-devel:

sudo yum install postgresql-devel


6) Install libaio package (should be already installed):

sudo yum libaio

---
[sudo] password for postgres:
Loaded plugins: langpacks, ulninfo
Repodata is over 2 weeks old. Install yum-cron? Or run: yum makecache fast
epel/x86_64/metalink                                                                                                  |  30 kB  00:00:00
epel                                                                                                                  | 4.7 kB  00:00:00
ol7_UEKR3                                                                                                             | 1.2 kB  00:00:00
ol7_latest                                                                                                            | 1.4 kB  00:00:00
pgdg10                                                                                                                | 3.6 kB  00:00:00
(1/4): epel/x86_64/group_gz                                                                                           |  88 kB  00:00:00
(2/4): pgdg10/7Server/x86_64/primary_db                                                                               | 227 kB  00:00:00
(3/4): epel/x86_64/updateinfo                                                                                         | 1.0 MB  00:00:01
(4/4): epel/x86_64/primary_db                                                                                         | 6.7 MB  00:00:01
(1/5): ol7_UEKR3/x86_64/updateinfo                                                                                    | 106 kB  00:00:00
(2/5): ol7_latest/x86_64/group                                                                                        | 810 kB  00:00:00
(3/5): ol7_latest/x86_64/updateinfo                                                                                   | 926 kB  00:00:00
(4/5): ol7_latest/x86_64/primary                                                                                      |  14 MB  00:00:05
(5/5): ol7_UEKR3/x86_64/primary                                                                                       |  38 MB  00:00:08
ol7_UEKR3                                                                                                                            885/885
ol7_latest                                                                                                                       12948/12948
Package libaio-0.3.109-13.el7.x86_64 already installed and latest version
Nothing to do
---


7) Set environment variables:

-> Oracle Instant Client libraries location:
/home/postgres/oracle_fdw/oracle-client/instantclient_19_3

-> Oracle Home:
/home/postgres/oracle_fdw/oracle-client/instantclient_19_3


export ORACLE_HOME=/home/postgres/oracle_fdw/oracle-client/instantclient_19_3
export LD_LIBRARY_PATH=/home/postgres/oracle_fdw/oracle-client/instantclient_19_3


8) Do the build:

-> Go to the location of the unzipped oracle_fdw:

cd /home/postgres/oracle_fdw/oracle_fdw-2.1.0


-- Eventually (to clean a current installation): make clean

-- Can be done with the postgres user:
make

---
[postgres@oel764-pg10-s1 oracle_fdw-2.1.0]$ make
gcc -Wall -Wmissing-prototypes -Wpointer-arith -Wdeclaration-after-statement -Wendif-labels -Wmissing-format-attribute -Wformat-security -fno-strict-aliasing -fwrapv -fexcess-precision=standard -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches -m64 -mtune=generic -fPIC -I/home/postgres/oracle_fdw/oracle-client/instantclient_19_3/sdk/include -I/home/postgres/oracle_fdw/oracle-client/instantclient_19_3/oci/include -I/home/postgres/oracle_fdw/oracle-client/instantclient_19_3/rdbms/public -I/usr/include/oracle/12.2/client -I/usr/include/oracle/12.2/client64 -I/usr/include/oracle/12.1/client -I/usr/include/oracle/12.1/client64 -I/usr/include/oracle/11.2/client -I/usr/include/oracle/11.2/client64 -I/usr/include/oracle/11.1/client -I/usr/include/oracle/11.1/client64 -I/usr/include/oracle/10.2.0.5/client -I/usr/include/oracle/10.2.0.5/client64 -I/usr/include/oracle/10.2.0.4/client -I/usr/include/oracle/10.2.0.4/client64 -I/usr/include/oracle/10.2.0.3/client -I/usr/include/oracle/10.2.0.3/client64 -I. -I./ -I/usr/pgsql-10/include/server -I/usr/pgsql-10/include/internal  -D_GNU_SOURCE -I/usr/include/libxml2  -I/usr/include  -c -o oracle_fdw.o oracle_fdw.c
gcc -Wall -Wmissing-prototypes -Wpointer-arith -Wdeclaration-after-statement -Wendif-labels -Wmissing-format-attribute -Wformat-security -fno-strict-aliasing -fwrapv -fexcess-precision=standard -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches -m64 -mtune=generic -fPIC -I/home/postgres/oracle_fdw/oracle-client/instantclient_19_3/sdk/include -I/home/postgres/oracle_fdw/oracle-client/instantclient_19_3/oci/include -I/home/postgres/oracle_fdw/oracle-client/instantclient_19_3/rdbms/public -I/usr/include/oracle/12.2/client -I/usr/include/oracle/12.2/client64 -I/usr/include/oracle/12.1/client -I/usr/include/oracle/12.1/client64 -I/usr/include/oracle/11.2/client -I/usr/include/oracle/11.2/client64 -I/usr/include/oracle/11.1/client -I/usr/include/oracle/11.1/client64 -I/usr/include/oracle/10.2.0.5/client -I/usr/include/oracle/10.2.0.5/client64 -I/usr/include/oracle/10.2.0.4/client -I/usr/include/oracle/10.2.0.4/client64 -I/usr/include/oracle/10.2.0.3/client -I/usr/include/oracle/10.2.0.3/client64 -I. -I./ -I/usr/pgsql-10/include/server -I/usr/pgsql-10/include/internal  -D_GNU_SOURCE -I/usr/include/libxml2  -I/usr/include  -c -o oracle_utils.o oracle_utils.c
gcc -Wall -Wmissing-prototypes -Wpointer-arith -Wdeclaration-after-statement -Wendif-labels -Wmissing-format-attribute -Wformat-security -fno-strict-aliasing -fwrapv -fexcess-precision=standard -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches -m64 -mtune=generic -fPIC -I/home/postgres/oracle_fdw/oracle-client/instantclient_19_3/sdk/include -I/home/postgres/oracle_fdw/oracle-client/instantclient_19_3/oci/include -I/home/postgres/oracle_fdw/oracle-client/instantclient_19_3/rdbms/public -I/usr/include/oracle/12.2/client -I/usr/include/oracle/12.2/client64 -I/usr/include/oracle/12.1/client -I/usr/include/oracle/12.1/client64 -I/usr/include/oracle/11.2/client -I/usr/include/oracle/11.2/client64 -I/usr/include/oracle/11.1/client -I/usr/include/oracle/11.1/client64 -I/usr/include/oracle/10.2.0.5/client -I/usr/include/oracle/10.2.0.5/client64 -I/usr/include/oracle/10.2.0.4/client -I/usr/include/oracle/10.2.0.4/client64 -I/usr/include/oracle/10.2.0.3/client -I/usr/include/oracle/10.2.0.3/client64 -I. -I./ -I/usr/pgsql-10/include/server -I/usr/pgsql-10/include/internal  -D_GNU_SOURCE -I/usr/include/libxml2  -I/usr/include  -c -o oracle_gis.o oracle_gis.c
gcc -Wall -Wmissing-prototypes -Wpointer-arith -Wdeclaration-after-statement -Wendif-labels -Wmissing-format-attribute -Wformat-security -fno-strict-aliasing -fwrapv -fexcess-precision=standard -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches -m64 -mtune=generic -fPIC -shared -o oracle_fdw.so oracle_fdw.o oracle_utils.o oracle_gis.o -L/usr/pgsql-10/lib -Wl,--as-needed  -L/usr/lib64 -Wl,--as-needed -Wl,-rpath,'/usr/pgsql-10/lib',--enable-new-dtags  -L/home/postgres/oracle_fdw/oracle-client/instantclient_19_3 -L/home/postgres/oracle_fdw/oracle-client/instantclient_19_3/bin -L/home/postgres/oracle_fdw/oracle-client/instantclient_19_3/lib -lclntsh -L/usr/lib/oracle/12.2/client/lib -L/usr/lib/oracle/12.2/client64/lib -L/usr/lib/oracle/12.1/client/lib -L/usr/lib/oracle/12.1/client64/lib -L/usr/lib/oracle/11.2/client/lib -L/usr/lib/oracle/11.2/client64/lib -L/usr/lib/oracle/11.1/client/lib -L/usr/lib/oracle/11.1/client64/lib -L/usr/lib/oracle/10.2.0.5/client/lib -L/usr/lib/oracle/10.2.0.5/client64/lib -L/usr/lib/oracle/10.2.0.4/client/lib -L/usr/lib/oracle/10.2.0.4/client64/lib -L/usr/lib/oracle/10.2.0.3/client/lib -L/usr/lib/oracle/10.2.0.3/client64/lib
[postgres@oel764-pg10-s1 oracle_fdw-2.1.0]$
[postgres@oel764-pg10-s1 oracle_fdw-2.1.0]$
---

-- Must be done as root or postgres (sudower):
sudo make install

---
[postgres@oel764-pg10-s1 oracle_fdw-2.1.0]$ sudo make install
[sudo] password for postgres:
/usr/bin/mkdir -p '/usr/lib64/pgsql'
/usr/bin/mkdir -p '/usr/share/pgsql/extension'
/usr/bin/mkdir -p '/usr/share/pgsql/extension'
/usr/bin/mkdir -p '/usr/share/doc/pgsql/extension'
/bin/sh /usr/lib64/pgsql/pgxs/src/makefiles/../../config/install-sh -c -m 755  oracle_fdw.so '/usr/lib64/pgsql/oracle_fdw.so'
/bin/sh /usr/lib64/pgsql/pgxs/src/makefiles/../../config/install-sh -c -m 644 ./oracle_fdw.control '/usr/share/pgsql/extension/'
/bin/sh /usr/lib64/pgsql/pgxs/src/makefiles/../../config/install-sh -c -m 644 ./oracle_fdw--1.1.sql ./oracle_fdw--1.0--1.1.sql  '/usr/share/pgsql/extension/'
/bin/sh /usr/lib64/pgsql/pgxs/src/makefiles/../../config/install-sh -c -m 644 ./README.oracle_fdw '/usr/share/doc/pgsql/extension/'
---

=> The extention seems to be created here:
/usr/share/pgsql/extension/

-> Location of the oracle_fdw extension:
/usr/share/pgsql/extension/

-> Three files are created:

---
-rw-r--r--. 1 root root   133 May 24 22:50 oracle_fdw.control
-rw-r--r--. 1 root root  1003 May 24 22:50 oracle_fdw--1.1.sql
-rw-r--r--. 1 root root   231 May 24 22:50 oracle_fdw--1.0--1.1.sql
---

-> Location of the PostgreSQL libraries:

pg_config --libdir 

---
/usr/pgsql-10/lib/
---

-> Location of PostgreSQL extension makefile:

pg_config --pgxs

---
/usr/pgsql-10/lib/pgxs/src/makefiles/pgxs.mk
---

-> Location of PostgreSQL extensions (show location of architecture-independent support files):
/usr/pgsql-10/share/extension

pg_config --sharedir

---
/usr/pgsql-10/share
---


9) Creating links for PostgreSQL:


sudo ln -s /usr/share/pgsql/extension/oracle_fdw.control /usr/pgsql-10/share/extension/oracle_fdw.control

sudo ln -s /usr/share/pgsql/extension/oracle_fdw--1.1.sql /usr/pgsql-10/share/extension/oracle_fdw--1.1.sql

sudo ln -s /usr/share/pgsql/extension/oracle_fdw--1.0--1.1.sql /usr/pgsql-10/share/extension/oracle_fdw--1.0--1.1.sql

sudo ln -s /home/postgres/oracle_fdw/oracle_fdw-2.1.0/oracle_fdw.so /usr/pgsql-10/lib/oracle_fdw.so

-> We end up with these symbolic links:

cd /usr/pgsql-10/share/extension

---
...
lrwxrwxrwx. 1 root root      45 May 24 22:55 oracle_fdw.control -> /usr/share/pgsql/extension/oracle_fdw.control
lrwxrwxrwx. 1 root root      46 May 24 22:57 oracle_fdw--1.1.sql -> /usr/share/pgsql/extension/oracle_fdw--1.1.sql
lrwxrwxrwx. 1 root root      51 May 24 22:57 oracle_fdw--1.0--1.1.sql -> /usr/share/pgsql/extension/oracle_fdw--1.0--1
---


10) Create the extension:

-> At that stage we have an empty database calles test.

-> When I try to create the exension for this database as:

psql -U postgres tests -c "CREATE EXTENSION oracle_fdw"

-> I get the error:

---
/home/postgres/oracle_fdw/oracle-client/instantclient_19_3:/home/postgres/oracle_fdw/oracle_fdw-2.1.0
[postgres@oel764-pg10-s1 ~]$ psql -U postgres tests -c "CREATE EXTENSION oracle_fdw"
ERROR:  could not load library "/usr/pgsql-10/lib/oracle_fdw.so": libclntsh.so.19.1: cannot open shared object file: No such file or directory
[postgres@oel764-pg10-s1 ~]$
---

-> But I have the oracle_fdw.so library:

---
[postgres@oel764-pg10-s1 ~]$ ls -l /usr/pgsql-10/lib/oracle_fdw.so
lrwxrwxrwx. 1 root root 56 May 24 23:42 /usr/pgsql-10/lib/oracle_fdw.so -> /home/postgres/oracle_fdw/oracle_fdw-2.1.0/oracle_fdw.so
[postgres@oel764-pg10-s1 ~]$
---

Can you tell me what is wrong or missing?

Tnanks by advance for your feedback

**Comment by laurenz** [comment]
You have to make sure that the shared library path `LD_LIBRARY_PATH` is set in the environment *of the PostgreSQL server process*.

For that, you either have to add the directory with `libclntsh.so` to `/etc/ld.so.conf` and run `ldconfig`, or you have to modify the PostgreSQL startup process so that it sets `LD_LIBRARY_PATH` in the postmaster's environment.

See for example #161 for more.

**Comment by MarvinDBA** [comment]
I was unable to really find what the problem was, but it is now solved.
Although I had defined the vaiables ORACLE_HOME and LD_LIBRARY_PATH as well as running ldconfig, I stil had this error.
I restarted the physical server many times (it is a virtual server), as well as the PostgreSQL server.
The variables ORACLE_HOME and LD_LIBRARY_PATH need to be defiined the same way as during the configuration and the installation of the distribution.
The command ldconfig needs to be run only once, as long as the two above variables are properly defined.
